File: app/src/main/java/at/shockbytes/corey/body/GoogleFitBodyManager.java
package at.shockbytes.corey.body;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentActivity;
import android.widget.Toast;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.Scopes;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.common.api.ResultCallback;
import com.google.android.gms.common.api.Scope;
import com.google.android.gms.fitness.Fitness;
import com.google.android.gms.fitness.data.DataPoint;
import com.google.android.gms.fitness.data.DataSet;
import com.google.android.gms.fitness.data.DataType;
import com.google.android.gms.fitness.data.Field;
import com.google.android.gms.fitness.request.DataReadRequest;
import com.google.android.gms.fitness.result.DataReadResult;

import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import at.shockbytes.corey.body.goal.Goal;
import at.shockbytes.corey.storage.StorageManager;
import at.shockbytes.corey.body.points.BodyFatPoint;
import at.shockbytes.corey.body.points.WeightPoint;
import at.shockbytes.corey.storage.live.LiveBodyUpdateListener;
import io.realm.RealmList;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import static at.shockbytes.corey.common.core.util.ResourceManager.roundDoubleWithDigits;

/**
 * @author Martin Macheiner
 *         Date: 04.08.2016.
 */
public class GoogleFitBodyManager implements BodyManager, ResultCallback<DataReadResult>,
        GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private Context context;
    private GoogleApiClient apiClient;

    private StorageManager storageManager;

    @Inject
    public GoogleFitBodyManager(Context context, StorageManager storageManager) {
        this.context = context;
        this.storageManager = storageManager;
    }

    @Override
    public void onResult(@NonNull DataReadResult res) {

        BodyInfo bodyInfo = new BodyInfo();
        bodyInfo.setDreamWeight(getDesiredWeight());
        bodyInfo.setWeightPoints(getWeightAsList(res.getDataSet(DataType.TYPE_WEIGHT)));
        bodyInfo.setHeight(getHeight(res.getDataSet(DataType.TYPE_HEIGHT)));
        bodyInfo.setBodyFatPoints(getBodyFatAsList(res.getDataSet(DataType.TYPE_BODY_FAT_PERCENTAGE)));

        // Store loaded bodyInfo, this will overwrite old value and #getBodyInfo() will always read
        // from local storage
        storageManager.appendToLocalBodyInfo(bodyInfo);
    }

    @Override
    public void poke(FragmentActivity activity) {

        if (apiClient == null) {
            apiClient = new GoogleApiClient.Builder(context)
                    .addApi(Fitness.HISTORY_API)
                    .addApi(Fitness.RECORDING_API)
                    .addScope(new Scope(Scopes.FITNESS_BODY_READ_WRITE))
                    .addScope(new Scope(Scopes.FITNESS_ACTIVITY_READ_WRITE))
                    .addConnectionCallbacks(this)
                    .enableAutoManage(activity, 0, this)
                    .build();
        }
    }

    @Override
    public Observable<BodyInfo> getBodyInfo() {
        return storageManager.getLocalBodyInfo()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io());
    }

    @Override
    public int getDesiredWeight() {
        return storageManager.getDesiredWeight();
    }

    @Override
    public void setDesiredWeight(int desiredWeight) {
        storageManager.setDesiredWeight(desiredWeight);
    }

    @Override
    public String getWeightUnit() {
        return storageManager.getWeightUnit();
    }

    @Override
    public void setWeightUnit(String unit) {
        storageManager.setWeightUnit(unit);
    }

    @Override
    public Observable<List<Goal>> getBodyGoals() {
        return storageManager.getBodyGoals()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.computation());
    }

    @Override
    public void updateBodyGoal(Goal g) {
        storageManager.updateBodyGoal(g);
    }

    @Override
    public void removeBodyGoal(Goal g) {
        storageManager.removeBodyGoal(g);
    }

    @Override
    public void storeBodyGoal(Goal g) {
        storageManager.storeBodyGoal(g);
    }

    @Override
    public void registerLiveBodyUpdates(LiveBodyUpdateListener listener) {
        storageManager.registerLiveBodyUpdates(listener);
    }

    @Override
    public void unregisterLiveBodyUpdates(LiveBodyUpdateListener listener) {
        storageManager.unregisterLiveBodyUpdates(listener);
    }

    @Override
    public void onConnected(@Nullable Bundle bundle) {
        loadFitnessData();
    }

    @Override
    public void onConnectionSuspended(int i) {
        Toast.makeText(context, "Connection suspended: " + i,
                Toast.LENGTH_LONG).show();
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        Toast.makeText(context,
                "Exception while connecting to Google Play Services: "
                        + connectionResult.getErrorMessage(),
                Toast.LENGTH_LONG).show();
    }

    private RealmList<WeightPoint> getWeightAsList(DataSet set) {

        RealmList<WeightPoint> weightPoints = new RealmList<>();
        for (DataPoint dp : set.getDataPoints()) {
            long time = dp.getStartTime(TimeUnit.MILLISECONDS);
            double weight = 0;
            for (Field field : dp.getDataType().getFields()) {
                weight = dp.getValue(field).asFloat();
            }
            weightPoints.add(new WeightPoint(time, roundDoubleWithDigits(weight, 1)));
        }
        return weightPoints;
    }

    private RealmList<BodyFatPoint> getBodyFatAsList(DataSet set) {

        RealmList<BodyFatPoint> bodyFatPoints = new RealmList<>();
        for (DataPoint dp : set.getDataPoints()) {
            long time = dp.getStartTime(TimeUnit.MILLISECONDS);
            double bodyFat = 0;
            for (Field field : dp.getDataType().getFields()) {
                bodyFat = dp.getValue(field).asFloat();
            }
            bodyFatPoints.add(new BodyFatPoint(time, roundDoubleWithDigits(bodyFat, 1)));
        }
        return bodyFatPoints;
    }

    private double getHeight(DataSet set) {

        double height = 0;
        if (set.getDataPoints() != null && set.getDataPoints().size() > 0) {
            int size = set.getDataPoints().size();
            DataPoint dp = set.getDataPoints().get(size - 1);
            for (Field field : dp.getDataType().getFields()) {
                height = roundDoubleWithDigits(dp.getValue(field).asFloat(), 2);
            }
        }
        return height;
    }

    private DataReadRequest buildGoogleFitRequest() {

        long startMillis = storageManager.getLastBodyInfoPull();
        //long startMillis = 1;
        long endMillis = System.currentTimeMillis();

        DataReadRequest request = new DataReadRequest.Builder()
                .setTimeRange(startMillis, endMillis, TimeUnit.MILLISECONDS)
                .read(DataType.TYPE_WEIGHT)
                .read(DataType.TYPE_HEIGHT)
                //.read(DataType.TYPE_BODY_FAT_PERCENTAGE)
                .build();

        // Store the latest pull timestamp
        storageManager.setLatestBodyInfoPull(endMillis);

        return request;
    }

    private void loadFitnessData() {

        DataReadRequest request = buildGoogleFitRequest();
        Fitness.HistoryApi.readData(apiClient, request).setResultCallback(this, 1, TimeUnit.MINUTES);
    }

}

File: app/src/main/java/at/shockbytes/corey/body/GoogleFitBodyManager.kt
package at.shockbytes.corey.body

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.widget.Toast
import at.shockbytes.corey.R
import at.shockbytes.corey.body.goal.Goal
import at.shockbytes.corey.body.info.BodyInfo
import at.shockbytes.corey.body.info.WeightPoint
import at.shockbytes.util.AppUtils
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.Scopes
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.Scope
import com.google.android.gms.fitness.Fitness
import com.google.android.gms.fitness.data.DataSet
import com.google.android.gms.fitness.data.DataType
import com.google.android.gms.fitness.request.DataReadRequest
import com.google.firebase.database.*
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit


/**
 * @author  Martin Macheiner
 * Date:    04.08.2016
 */
class GoogleFitBodyManager(private val context: Context,
                           private val preferences: SharedPreferences,
                           private val firebase: FirebaseDatabase) : BodyManager,
        GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    init {
        setupFirebase()
    }

    private var apiClient: GoogleApiClient? = null
    private var _bodyInfo: BodyInfo = BodyInfo()

    private var _desiredWeight: Int = -1
    private var _goals: MutableList<Goal> = mutableListOf()

    private var bodyListener: LiveBodyUpdateListener? = null

    override var desiredWeight: Int
        get() {
            // No sync with firebase, use local cached value
            return if (_desiredWeight < 0) {
                preferences.getInt(PREF_DREAM_WEIGHT, 0)
            } else _desiredWeight
        }
        set(value) {
            preferences.edit().putInt(PREF_DREAM_WEIGHT, value).apply()
            firebase.getReference("/body/desired").setValue(value)
        }

    override val weightUnit: String
        get() = preferences.getString(PREF_WEIGHT_UNIT, context.getString(R.string.default_weight_unit))
                ?: context.getString(R.string.default_weight_unit)

    override val bodyGoals: Observable<List<Goal>>
        get() = Observable.just(_goals.toList())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.computation())

    override val bodyInfo: Observable<BodyInfo>
        get() {
            return if (_bodyInfo.isNotEmpty) {
                Observable.just(_bodyInfo)
            } else {
                loadFitnessDataSync() // Request new data if nothing is available
            }.observeOn(AndroidSchedulers.mainThread()).subscribeOn(Schedulers.io())
        }

    override fun poke(activity: FragmentActivity?) {
        if (apiClient == null && activity != null) {
            apiClient = GoogleApiClient.Builder(context)
                    .addApi(Fitness.HISTORY_API)
                    .addApi(Fitness.RECORDING_API)
                    .addScope(Scope(Scopes.FITNESS_BODY_READ_WRITE))
                    .addScope(Scope(Scopes.FITNESS_ACTIVITY_READ_WRITE))
                    .addConnectionCallbacks(this)
                    .enableAutoManage(activity, 0, this)
                    .build()
        }
    }

    override fun onConnected(bundle: Bundle?) {
        loadFitnessData()
    }

    override fun onConnectionSuspended(i: Int) {
        Toast.makeText(context, "Connection suspended: $i", Toast.LENGTH_LONG).show()
    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {
        Toast.makeText(context,
                "Exception while connecting to Google Play Services: ${connectionResult.errorMessage}",
                Toast.LENGTH_LONG).show()
    }

    override fun registerLiveBodyUpdates(listener: LiveBodyUpdateListener) {
        this.bodyListener = listener
    }

    override fun unregisterLiveBodyUpdates() {
        bodyListener = null
    }

    override fun updateBodyGoal(g: Goal) {
        firebase.getReference("/body/goal").child(g.id).setValue(g)
    }

    override fun removeBodyGoal(g: Goal) {
        firebase.getReference("/body/goal").child(g.id).removeValue()
    }

    override fun storeBodyGoal(g: Goal) {
        val ref = firebase.getReference("/body/goal").push()
        g.id = ref.key ?: ""
        ref.setValue(g)
    }

    // -----------------------------------------------------------------------------------------

    private fun loadFitnessData() {
        Fitness.HistoryApi.readData(apiClient, buildGoogleFitRequest())
                .setResultCallback { res ->
                    _bodyInfo = BodyInfo(getWeightList(res.getDataSet(DataType.TYPE_WEIGHT)),
                            getHeight(res.getDataSet(DataType.TYPE_HEIGHT)),
                            desiredWeight)
                }
    }

    private fun loadFitnessDataSync(): Observable<BodyInfo> {
        return Observable.fromCallable {
            val res = Fitness.HistoryApi.readData(apiClient, buildGoogleFitRequest()).await()
            _bodyInfo = BodyInfo(getWeightList(res.getDataSet(DataType.TYPE_WEIGHT)),
                    getHeight(res.getDataSet(DataType.TYPE_HEIGHT)),
                    desiredWeight)
            _bodyInfo
        }
    }

    private fun buildGoogleFitRequest(): DataReadRequest {

        val startMillis = 1.toLong()
        val endMillis = System.currentTimeMillis()

        return DataReadRequest.Builder()
                .setTimeRange(startMillis, endMillis, TimeUnit.MILLISECONDS)
                .read(DataType.TYPE_WEIGHT)
                .read(DataType.TYPE_HEIGHT)
                //.read(DataType.TYPE_BODY_FAT_PERCENTAGE)
                .build()
    }

    private fun getWeightList(set: DataSet): List<WeightPoint> {
        return set.dataPoints.mapTo(mutableListOf()) {
            val timeStamp = it.getStartTime(TimeUnit.MILLISECONDS)
            val weight = it.getValue(it.dataType.fields[0]).asFloat().toDouble()
            WeightPoint(timeStamp, AppUtils.roundDouble(weight, 1))
        }
    }

    private fun getHeight(set: DataSet): Double {
        return if (set.dataPoints.isNotEmpty()) {
            val dp = set.dataPoints[set.dataPoints.size - 1]
            AppUtils.roundDouble(dp.getValue(dp.dataType.fields[0]).asFloat().toDouble(), 2)
        } else 0.0
    }

    private fun setupFirebase() {

        firebase.getReference("/body/desired").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val data = dataSnapshot.value
                if (data != null) {
                    _desiredWeight = Integer.parseInt(data.toString())

                    preferences.edit().putInt(PREF_DREAM_WEIGHT, _desiredWeight).apply()
                    bodyListener?.onDesiredWeightChanged(_desiredWeight)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
            }
        })

        firebase.getReference("body/goal").addChildEventListener(object : ChildEventListener {
            override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {

                dataSnapshot.getValue(Goal::class.java)?.let { g ->
                    _goals.add(g)
                    bodyListener?.onBodyGoalAdded(g)
                }
            }

            override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {

                dataSnapshot.getValue(Goal::class.java)?.let { g ->
                    _goals[_goals.indexOf(g)] = g
                    bodyListener?.onBodyGoalChanged(g)
                }
            }

            override fun onChildRemoved(dataSnapshot: DataSnapshot) {

                dataSnapshot.getValue(Goal::class.java)?.let { g ->
                    _goals.remove(g)
                    bodyListener?.onBodyGoalDeleted(g)
                }
            }

            override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) = Unit

            override fun onCancelled(databaseError: DatabaseError) = Unit
        })

    }


    companion object {

        private const val PREF_DREAM_WEIGHT = "dreamweight"
        private const val PREF_WEIGHT_UNIT = "weight_unit"

    }

}
